src/riak_kv_vnode.erl

```
@@ -2367,12 +2380,31 @@ encode_and_put_no_sib_check(Obj, Mod, Bucket, Key, IndexSpecs, ModState,
                         false ->
                             ok
                     end,
-                    PutRet = Mod:put(Bucket, Key, IndexSpecs, EncodedVal,
-                                     ModState),
+                    case Sync of
+                        true ->
+                            PutRet = maybe_flush_put(Mod, Bucket, Key, IndexSpecs,
+                                                     EncodedVal, ModState);
+                        false ->
+                            PutRet = Mod:put(Bucket, Key, IndexSpecs, EncodedVal,
+                                             ModState)
+                    end,
                     {PutRet, EncodedVal}
             end
     end.
 
+-spec maybe_flush_put(Mod::term(), Bucket::riak_object:bucket(), Key::riak_object:key(),
+                      IndexSpecs::list(), EncodedVal::binary(), ModState::term()) ->
+           {{ok, UpdModState::term()}, EncodedObj::binary()} |
+           {{error, Reason::term(), UpdModState::term()}, EncodedObj::binary()}.
+maybe_flush_put(Mod, Bucket, Key, IndexSpecs, EncodedVal, ModState) ->
+    {ok, Capabilities} = Mod:capabilities(ModState),
+    case lists:member(flush_put, Capabilities) of
+        true ->
+            Mod:flush_put(Bucket, Key, IndexSpecs, EncodedVal, ModState);
+        false ->
```

Yeah, I'm not sure what to do here, maybe "roll a dice" and only log one in X? I don't know if lager has an option for sample logging. Or we can increment a stat, that's less spammy maybe, a counter maybe? Not sure myself, but I agree that silent failure/discarding of user intent is a bad idea.